src/utils.js

```diff
@@ -654,35 +654,54 @@ module.exports.inherits = function(ctor, superCtor) {
             configurable: true,
         },
     });
-};
+}
+
+/**
+ * Polyfills inheritance for prototypes by allowing different kinds of
+ * super types. Typically prototypes would use `SuperType.call(this, params)`
+ * though this doesn't always work in some environments - this function
+ * falls back to using `Object.assign()` to clone a constructed copy
+ * of the super type onto `thisArg`.
+ * @param {any} thisArg The child instance. Modified in place.
+ * @param {any} SuperType The type to act as a super instance
+ * @param {any} params Arguments to supply to the super type's constructor
+ */
+export function polyfillSuper(thisArg, SuperType, ...params) {
+    try {
+        SuperType.call(thisArg, ...params);
+    } catch (e) {
+        // fall back to Object.assign to just clone the thing
+        const fakeSuper = new SuperType(...params);
+        Object.assign(thisArg, fakeSuper);
+    }
+}
```

Comment:
This seems very surprising since it's what we were doing before? What's regressed such that we can no longer reliably do this and why are our code standards going backwards?

Reply:
This is also a result of ES6 vs older CommonJS syntax: something goes wrong with the prototypes that prevents stuff like `EncryptionAlgorithm` from being used as a base type in the traditional way, but only depending on who is calling the code. Prototype inheritance is strange and stuff like Jest expects all the code to be at the same standard and as such fails when we mix the two: `MegolmEncryption` for instance is a prototype that uses an `EncryptionAlgorithm` class as its base - we're mixing two concepts here, and the only reason it worked without issue before was because our code was transpiled aggressively to the point where the runtime couldn't see the types and just hoped for the best. Now that we are using a less aggressive transpilation stage, we start to run into problems in mixing concepts.

An alternative fix is to convert all of our prototypes to classes, following the standard set by the base classes, though that is much more involved than a polyfill. Using a polyfill makes this PR less invasive and still pass CI.

Reply:
OK. Perhaps what would help is some examples of environments where it does & doesn't work?

Reply:
Jest and webpack (using babel-loader with babel 7) have the issue at a minimum. The error is roughly "Cannot use EncryptionAlgorithm without the new keyword" (it wants you to use `new EncryptionAlgorithm(...)`). The browserified bundle doesn't seem to care, annoyingly.

Reply:
OK - I assume an ES6 style `super()` call doesn't work either? Slightly scary having different code paths being executed between the two different versions of the js-sdk but if it's what is necessary then I guess it's what we have to do. +1 to converting to real classes ASAP.

Reply:
it's a prototype so `super()` isn't a valid keyword :(

Reply:
ok - fair enough